src/cursor/cur_file.c

```diff
@@ -75,6 +75,103 @@ __curfile_equals(WT_CURSOR *a, WT_CURSOR *b, int *equalp)
     API_END_RET(session, ret);
 }
 
+/*
+ * __curfile_set_key --
+ *     WT_CURSOR->set_key implementation for file cursor.
+ */
+static void
+__curfile_set_key(WT_CURSOR *cursor, ...)
+{
+    va_list ap;
+
+    va_start(ap, cursor);
+    F_CLR((WT_CURSOR_BTREE *)cursor, WT_CBT_REPOSITION);
+    WT_IGNORE_RET(__wt_cursor_set_keyv(cursor, cursor->flags, ap));
+    va_end(ap);
+}
+
+/*
+ * __curfile_reposition_timing_stress --
+ *     Optionally reposition the cursor 10% of times
+ */
+static bool
+__curfile_reposition_timing_stress(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+
+    if (FLD_ISSET(conn->timing_stress_flags, WT_TIMING_STRESS_BTREE_REPOSITION) &&
+      __wt_random(&session->rnd) % 10 == 0)
+        return (true);
+
+    return (false);
+}
+
+/*
+ * __curfile_reposition --
+ *     Reposition the saved key
+ */
+static int
+__curfile_reposition(WT_CURSOR *cursor, bool exact_match, bool next, bool *moved)
+{
+    WT_CURSOR_BTREE *cbt;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    int exact;
+
+    cbt = (WT_CURSOR_BTREE *)cursor;
+    session = CUR2S(cursor);
+
+    WT_STAT_CONN_DATA_INCR(session, cursor_reposition);
+
+    if (moved != NULL)
+        *moved = false;
+
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_EXT))
+        WT_ERR(__wt_panic(session, EINVAL, "Reposition flag is set without a search key"));
+
+    WT_ERR(__wt_btcur_search_near(cbt, &exact));
+
+    if (!exact_match) {
+        if ((exact > 0 && next) || (exact < 0 && !next))
+            *moved = true;
+    } else if (exact != 0)
+        WT_ERR(WT_NOTFOUND);
+
+    /* Search maintains a position, key and value. */
+    WT_ASSERT(session,
+      F_ISSET(cbt, WT_CBT_ACTIVE) && F_MASK(cursor, WT_CURSTD_KEY_SET) == WT_CURSTD_KEY_INT &&
+        F_MASK(cursor, WT_CURSTD_VALUE_SET) == WT_CURSTD_VALUE_INT);
+
+err:
+    F_CLR(cbt, WT_CBT_REPOSITION);
+    if (ret != 0)
+        WT_STAT_CONN_DATA_INCR(session, cursor_reposition_failed);
+    return (ret);
+}
+
+/*
+ * __curfile_release_page --
+ *     Copy the return key value to the local buffer and release the page.
+ */
+static int
+__curfile_release_page(WT_CURSOR *cursor)
+{
+    WT_CURSOR_BTREE *cbt;
+
+    cbt = (WT_CURSOR_BTREE *)cursor;
+
+    WT_RET(__wt_cursor_localkey(cursor));
+    WT_RET(__wt_cursor_localvalue(cursor));
+    WT_RET(__wt_cursor_reset(cbt));
+    F_SET(cbt, WT_CBT_REPOSITION);
+
+    WT_STAT_CONN_DATA_INCR(CUR2S(cursor), cursor_release_page_to_reposition);
```

Comment:
I view the btree code as functional building blocks that provides basic features to search and move the btree cursor. Here we are adding a layer of abstraction above the btree layer. If we need to implement this layer of abstraction in the btree layer, we need to define new btree level functions on top of the current functions, such as __bt_cur_search. I don't see much value in that as we can do this just here in the cursor. But I agree we probably should move the flag to WT_CURSTD_XXX.

Reply:
I wouldn't do it with new, wrapping functions, I'd do it by hooking the standard functions exactly as you've done here.

Honestly, I'm torn: (1) I  think the btree layer "owns" cursor positioning, that's not something I expect higher-level functions to know about, and (2) this change is also going to constrain what btree cursors have this functionality, that is, only `file:` cursors get it. If you imagine something like LSM that calls the btree layer functions directly, where we put this functionality matters. (To be fair, there's nothing like that now.)

@agorrod, what do you think?

Reply:
I'll try to have another pr implementing all this in the btree layer.

Reply:
I'd wait -- @agorrod might just agree with you. :)

Reply:
I agree with @keithbostic - I think this would work better in the btree code.
I think the `Here we are adding a layer of abstraction above the btree layer.` position is an artifact of how we got to doing this, not necessarily an indication of how the implementation should work now that it's being pulled into WiredTiger.
The two things that make me agree with Keith are the WT_PAGE layering violation, and the repeated calls in all the different cursor functions.
OTOH: If the change doesn't work in the btree code, I think we can make this approach do what we need in a reasonable way.

Reply:
I'm looking at how to implement this in the btree layer. I think we definitely need a wrapper around the `__wt_btcur_search` function as other write functions depends on `__wt_btcur_search`. Therefore, we cannot add the feature that deposition the cursor in there. If we do it there, at least we should add a flag that indicates we should enable the feature.

Reply:
Without seeing the code, it's hard to weigh in with confidence, but a flag feels consistent with how we do similar things in other places.

Reply:
Do the other write functions not want the reposition effect?

Reply:
What I mean is that, for example, insert will do a search if the key is not positioned, however, if we do everything in search, it may release the position contradicting what we needs it to do. We may need to add a flag to the search function say don't release the cursor in this case.

Reply:
Yes, that makes sense to me, I had not thought of that question.

I don't know how many of these there are.

For example, `__wt_btcur_insert()` doesn't call `__wt_btcur_search()` (it calls the underlying tree search functions like `__cursor_col_search()`), and I'd expect the reposition code to be in `__wt_btcur_search()`, something like:
```
if (ret != 0 || F_ISSET(cbt.ref.page, WT_PAGE_FORCE_EVICT)) {
    WT_TRET(__wt_cursor_localkey(cursor));    // Maybe these calls get added to __cursor_reset();
    WT_TRET(__wt_cursor_localvalue(cursor));
    WT_TRET(__cursor_reset(cbt));
}
if (ret != 0)
    __cursor_state_restore(cursor, &state);
```

I won't speak for @agorrod, but I'm still on the fence on this one -- I'm ready to hear that you took a look and this is insane. :)

Reply:
@keithbostic I think it is achievable to implement this in btree. But some asserts at the cursor level need to change as well. I'm going on holiday from next week and someone else should take over this work. I'll let him try with the btree solution.

Reply:
@quchenhao, sounds good, have a great vacation!

Let me know if you don't find anybody else to take a run at this one, I can try if you want.